{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28d7ecb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import kipoiseq\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import torch\n",
    "import torch.nn.functional as F\n",
    "from train import TrainModule, FastaStringExtractor\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "import pyBigWig\n",
    "from kipoiseq import Interval\n",
    "import math\n",
    "import random\n",
    "from scipy.stats import pearsonr\n",
    "\n",
    "random.seed(2077)\n",
    "SEQUENCE_LENGTH = 65536\n",
    "\n",
    "def one_hot_encode(sequence):\n",
    "    #return kipoiseq.transforms.functional.one_hot_dna(sequence, alphabet=('A', 'C', 'G', 'T'),).astype(np.float32)\n",
    "    #en_dict = {'A' : 0, 'C' : 1, 'G' : 2, 'T' : 3, 'N' : 4}\n",
    "    en_dict = {'A' : 0, 'T' : 1, 'C' : 2, 'G' : 3, 'N' : 4}\n",
    "    en_seq = [en_dict[ch] for ch in sequence]\n",
    "    np_seq = np.array(en_seq, dtype = int)\n",
    "    seq_emb = np.zeros((len(np_seq), 5))\n",
    "    seq_emb[np.arange(len(np_seq)), np_seq] = 1\n",
    "    #seq_emb = np.eye(5)[np_seq]\n",
    "    return seq_emb.astype(np.float32)\n",
    "\n",
    "def plot_tracks(tracks, interval, height=1.5):\n",
    "    fig, axes = plt.subplots(len(tracks), 1, figsize=(20, height * len(tracks)), sharex=True)\n",
    "    for ax, (title, y) in zip(axes, tracks.items()):\n",
    "        ax.fill_between(np.linspace(interval.start, interval.end, num=len(y)), y)\n",
    "        ax.set_title(title)\n",
    "        sns.despine(top=True, right=True, bottom=True)\n",
    "    ax.set_xlabel(str(interval))\n",
    "    plt.tight_layout()\n",
    "    \n",
    "def variant_generator(vcf_file):\n",
    "    with open(vcf_file) as f:\n",
    "        for line in f:\n",
    "            chrom, pos, id, ref, alt = line.split('\\t')[:5]\n",
    "            \n",
    "      # Split ALT alleles and return individual variants as output.\n",
    "            yield kipoiseq.Variant(chrom=chrom, pos=pos,ref=ref, alt=alt, id=id)\n",
    "        \n",
    "def get_cds_interval(gff_df, gene_name):\n",
    "    # filter by gene_name\n",
    "    gff_df = gff_df[gff_df[8].str.split(';').str[5].str.split('=').str[1] == gene_name]\n",
    "\n",
    "    # generate dict\n",
    "    gene_dict = {row[8].split(';')[0].split('=')[1]: {'chrom': row[0],'start': row[3], 'end': row[4], 'name':row[8].split(';')[7].split('=')[1], \n",
    "                                                      'cds_intervals': []} for _, row in gff_df.iterrows() if row[2] == 'transcript'}\n",
    "    cds_df = gff_df[gff_df[2] == 'CDS']\n",
    "    for _, row in cds_df.iterrows():\n",
    "        gene_name = row[8].split(';')[3].split('=')[1]\n",
    "        gene_dict[gene_name]['cds_intervals'].append((row[3], row[4]))\n",
    "    target_gene = next((gene_info for gene_info in gene_dict.values()\n",
    "                      ), None)\n",
    "    if target_gene:\n",
    "        cds_intervals = target_gene['cds_intervals']\n",
    "        return {\n",
    "            'chrom': target_gene['chrom'],\n",
    "            'gene_start': target_gene['start'],\n",
    "            'gene_end': target_gene['end'],\n",
    "            'cds_intervals': cds_intervals,\n",
    "            'gene_name': target_gene['name']\n",
    "        }\n",
    "    \n",
    "def generate_inputs(region, fasta_file, bw_file, region_len=SEQUENCE_LENGTH):\n",
    "    bw = pyBigWig.open(bw_file)\n",
    "    target = []\n",
    "    chrom = region.chrom\n",
    "    start = region.start\n",
    "    end = region.end\n",
    "    chromosome_length = bw.chroms(chrom)\n",
    "    interval = Interval(chrom, start, end).resize(region_len)\n",
    "    trimmed_interval = Interval(interval.chrom,\n",
    "                                max(interval.start, 0),\n",
    "                                min(interval.end, chromosome_length),\n",
    "                                )\n",
    "    signals = np.array(bw.values(chrom, trimmed_interval.start, trimmed_interval.end)).astype(np.float32).tolist()\n",
    "    pad_upstream = np.array([0] * max(-interval.start, 0)).astype(np.float32).tolist()\n",
    "    pad_downstream = np.array([0] * max(interval.end - chromosome_length, 0)).astype(np.float32).tolist()\n",
    "    tmp = pad_upstream + signals + pad_downstream\n",
    "    arr = np.array(tmp).astype(np.float32)\n",
    "    target.append(arr)\n",
    "\n",
    "    target = np.array(target).astype(np.float32)\n",
    "    target = np.nan_to_num(target,0)\n",
    "    target = np.log(target + 1)\n",
    "    bw.close()\n",
    "    return target\n",
    "\n",
    "def generate_outputs(region, fasta_file, bw_file, nBins=1024, region_len=65536):\n",
    "    bw = pyBigWig.open(bw_file)\n",
    "    target = []\n",
    "    chrom = region.chrom\n",
    "    start = region.start\n",
    "    end = region.end\n",
    "    chromosome_length = bw.chroms(chrom)\n",
    "    interval = Interval(chrom, start, end).resize(region_len)\n",
    "    trimmed_interval = Interval(interval.chrom,\n",
    "                                max(interval.start, 0),\n",
    "                                min(interval.end, chromosome_length),\n",
    "                               )\n",
    "    signals = np.array(bw.values(chrom, trimmed_interval.start, trimmed_interval.end)).astype(np.float32).tolist()\n",
    "    pad_upstream = np.array([0] * max(-interval.start, 0)).astype(np.float32).tolist()\n",
    "    pad_downstream = np.array([0] * max(interval.end - chromosome_length, 0)).astype(np.float32).tolist()\n",
    "    tmp = pad_upstream + signals + pad_downstream\n",
    "    arr = np.array(tmp).astype(np.float32)\n",
    "    reshaped_arr = arr.reshape(-1, 64)\n",
    "    averages = np.mean(reshaped_arr, axis=1)\n",
    "    target.append(averages)\n",
    "\n",
    "    target = np.array(target).astype(np.float32)\n",
    "    target = np.nan_to_num(target,0)\n",
    "    target = np.log(target + 1)\n",
    "    bw.close()\n",
    "    return target\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "device = 'cuda:0'\n",
    "checkpoint = '/data/slurm/hejl/riboseq/results_DNA/bigmodel/bigmodel_h512_l12_lr1e-5/models/epoch=38-step=746889.ckpt'\n",
    "model = TrainModule.load_from_checkpoint(checkpoint).to(device)\n",
    "model = model.eval()\n",
    "\n",
    "#GTF_FILE = '/data/slurm/leixiong/m6A_prediction/data/gencode.v42.chr_patch_hapl_scaff.annotation.gtf.gz'\n",
    "SEQUENCE_LENGTH = 65536\n",
    "fasta_file = '/data/slurm/hejl/riboseq/data/hg38/hg38.fa'#str(DATA_PATH / 'hg38.fa')\n",
    "fasta_extractor = FastaStringExtractor(fasta_file)\n",
    "gff_file = '/data/slurm/hejl/riboseq/gencode.v43.annotation.gff3'\n",
    "#rna_bw_file = '/data/slurm/hejl/riboseq/data/hg38/K562/GSE153597/input_features/rnaseq.bw'\n",
    "rna_bw_file = '/data/slurm/hejl/riboseq/data/hg38/neuron/GSE90469_2/input_features/rnaseq.bw'\n",
    "ribo_bw_file = '/data/slurm/hejl/riboseq/data/hg38/neuron/GSE90469_2/output_features/riboseq.bw'\n",
    "\n",
    "gene_name = 'OR4F29'\n",
    "#gene_name = 'KRT19'\n",
    "\n",
    "gff_df = pd.read_csv(gff_file, sep='\\t', comment='#', header=None)\n",
    "gene_info = get_cds_interval(gff_df, gene_name)\n",
    "\n",
    "gene_interval =kipoiseq.Interval(gene_info['chrom'], gene_info['gene_start'], gene_info['gene_end'])\n",
    "region_interval = gene_interval\n",
    "target_interval = Interval(region_interval.chrom, region_interval.start, region_interval.end).resize(SEQUENCE_LENGTH)\n",
    "ref_seq = fasta_extractor.extract(target_interval)\n",
    "ref_emb = torch.Tensor(one_hot_encode(ref_seq)).to(device)\n",
    "\n",
    "epi = torch.Tensor(generate_inputs(region_interval, fasta_file, rna_bw_file)[0]).unsqueeze(1).to(device)\n",
    "reference_input = torch.cat([ref_emb, epi], dim = 1).unsqueeze(0)\n",
    "pred = model(reference_input)[0].detach().cpu().numpy()\n",
    "\n",
    "gt = torch.Tensor(generate_outputs(region_interval, fasta_file, ribo_bw_file)[0]).cpu().numpy()\n",
    "#\n",
    "tracks = {'Ground Truth(Epithelial)': gt,\n",
    "          'Prediction(Epithelial)': pred\n",
    "          }\n",
    "\n",
    "plot_tracks(tracks, target_interval, height=1) #\n",
    "\n",
    "correlation, p_value = pearsonr(pred,gt)\n",
    "\n",
    "print('Pearson: ')\n",
    "print(correlation)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "387b377b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "77878450",
   "metadata": {},
   "source": [
    "##"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f185524d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import kipoiseq\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import torch\n",
    "import torch.nn.functional as F\n",
    "from train import TrainModule, FastaStringExtractor\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "import pyBigWig\n",
    "from kipoiseq import Interval\n",
    "import math\n",
    "import random\n",
    "from scipy.stats import pearsonr\n",
    "\n",
    "random.seed(2077)\n",
    "SEQUENCE_LENGTH = 65536\n",
    "\n",
    "def one_hot_encode(sequence):\n",
    "    #return kipoiseq.transforms.functional.one_hot_dna(sequence, alphabet=('A', 'C', 'G', 'T'),).astype(np.float32)\n",
    "    #en_dict = {'A' : 0, 'C' : 1, 'G' : 2, 'T' : 3, 'N' : 4}\n",
    "    en_dict = {'A' : 0, 'T' : 1, 'C' : 2, 'G' : 3, 'N' : 4}\n",
    "    en_seq = [en_dict[ch] for ch in sequence]\n",
    "    np_seq = np.array(en_seq, dtype = int)\n",
    "    seq_emb = np.zeros((len(np_seq), 5))\n",
    "    seq_emb[np.arange(len(np_seq)), np_seq] = 1\n",
    "    #seq_emb = np.eye(5)[np_seq]\n",
    "    return seq_emb.astype(np.float32)\n",
    "\n",
    "def plot_tracks(tracks, interval, height=1.5):\n",
    "    fig, axes = plt.subplots(len(tracks), 1, figsize=(20, height * len(tracks)), sharex=True)\n",
    "    for ax, (title, y) in zip(axes, tracks.items()):\n",
    "        ax.fill_between(np.linspace(interval.start, interval.end, num=len(y)), y)\n",
    "        ax.set_title(title)\n",
    "        sns.despine(top=True, right=True, bottom=True)\n",
    "    ax.set_xlabel(str(interval))\n",
    "    plt.tight_layout()\n",
    "    \n",
    "def variant_generator(vcf_file):\n",
    "    with open(vcf_file) as f:\n",
    "        for line in f:\n",
    "            chrom, pos, id, ref, alt = line.split('\\t')[:5]\n",
    "            \n",
    "      # Split ALT alleles and return individual variants as output.\n",
    "            yield kipoiseq.Variant(chrom=chrom, pos=pos,ref=ref, alt=alt, id=id)\n",
    "        \n",
    "def get_cds_interval(gff_df, gene_name):\n",
    "    # filter by gene_name\n",
    "    gff_df = gff_df[gff_df[8].str.split(';').str[5].str.split('=').str[1] == gene_name]\n",
    "\n",
    "    # generate dict\n",
    "    gene_dict = {row[8].split(';')[0].split('=')[1]: {'chrom': row[0],'start': row[3], 'end': row[4], 'name':row[8].split(';')[7].split('=')[1], \n",
    "                                                      'cds_intervals': []} for _, row in gff_df.iterrows() if row[2] == 'transcript'}\n",
    "    cds_df = gff_df[gff_df[2] == 'CDS']\n",
    "    for _, row in cds_df.iterrows():\n",
    "        gene_name = row[8].split(';')[3].split('=')[1]\n",
    "        gene_dict[gene_name]['cds_intervals'].append((row[3], row[4]))\n",
    "    target_gene = next((gene_info for gene_info in gene_dict.values()\n",
    "                      ), None)\n",
    "    if target_gene:\n",
    "        cds_intervals = target_gene['cds_intervals']\n",
    "        return {\n",
    "            'chrom': target_gene['chrom'],\n",
    "            'gene_start': target_gene['start'],\n",
    "            'gene_end': target_gene['end'],\n",
    "            'cds_intervals': cds_intervals,\n",
    "            'gene_name': target_gene['name']\n",
    "        }\n",
    "    \n",
    "def generate_inputs(region, fasta_file, bw_file, region_len=SEQUENCE_LENGTH):\n",
    "    bw = pyBigWig.open(bw_file)\n",
    "    target = []\n",
    "    chrom = region.chrom\n",
    "    start = region.start\n",
    "    end = region.end\n",
    "    chromosome_length = bw.chroms(chrom)\n",
    "    interval = Interval(chrom, start, end).resize(region_len)\n",
    "    trimmed_interval = Interval(interval.chrom,\n",
    "                                max(interval.start, 0),\n",
    "                                min(interval.end, chromosome_length),\n",
    "                                )\n",
    "    signals = np.array(bw.values(chrom, trimmed_interval.start, trimmed_interval.end)).astype(np.float32).tolist()\n",
    "    pad_upstream = np.array([0] * max(-interval.start, 0)).astype(np.float32).tolist()\n",
    "    pad_downstream = np.array([0] * max(interval.end - chromosome_length, 0)).astype(np.float32).tolist()\n",
    "    tmp = pad_upstream + signals + pad_downstream\n",
    "    arr = np.array(tmp).astype(np.float32)\n",
    "    target.append(arr)\n",
    "\n",
    "    target = np.array(target).astype(np.float32)\n",
    "    target = np.nan_to_num(target,0)\n",
    "    target = np.log(target + 1)\n",
    "    bw.close()\n",
    "    return target\n",
    "\n",
    "def generate_outputs(region, fasta_file, bw_file, nBins=1024, region_len=65536):\n",
    "    bw = pyBigWig.open(bw_file)\n",
    "    target = []\n",
    "    chrom = region.chrom\n",
    "    start = region.start\n",
    "    end = region.end\n",
    "    chromosome_length = bw.chroms(chrom)\n",
    "    interval = Interval(chrom, start, end).resize(region_len)\n",
    "    trimmed_interval = Interval(interval.chrom,\n",
    "                                max(interval.start, 0),\n",
    "                                min(interval.end, chromosome_length),\n",
    "                               )\n",
    "    signals = np.array(bw.values(chrom, trimmed_interval.start, trimmed_interval.end)).astype(np.float32).tolist()\n",
    "    pad_upstream = np.array([0] * max(-interval.start, 0)).astype(np.float32).tolist()\n",
    "    pad_downstream = np.array([0] * max(interval.end - chromosome_length, 0)).astype(np.float32).tolist()\n",
    "    tmp = pad_upstream + signals + pad_downstream\n",
    "    arr = np.array(tmp).astype(np.float32)\n",
    "    reshaped_arr = arr.reshape(-1, 64)\n",
    "    averages = np.mean(reshaped_arr, axis=1)\n",
    "    target.append(averages)\n",
    "\n",
    "    target = np.array(target).astype(np.float32)\n",
    "    target = np.nan_to_num(target,0)\n",
    "    target = np.log(target + 1)\n",
    "    bw.close()\n",
    "    return target\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "device = 'cuda:0'\n",
    "checkpoint = '/data/slurm/hejl/riboseq/results_DNA/bigmodel/bigmodel_h512_l12_lr1e-5/models/epoch=38-step=746889.ckpt'\n",
    "model = TrainModule.load_from_checkpoint(checkpoint).to(device)\n",
    "model = model.eval()\n",
    "\n",
    "#GTF_FILE = '/data/slurm/leixiong/m6A_prediction/data/gencode.v42.chr_patch_hapl_scaff.annotation.gtf.gz'\n",
    "SEQUENCE_LENGTH = 65536\n",
    "fasta_file = '/data/slurm/hejl/riboseq/data/hg38/hg38.fa'#str(DATA_PATH / 'hg38.fa')\n",
    "fasta_extractor = FastaStringExtractor(fasta_file)\n",
    "gff_file = '/data/slurm/hejl/riboseq/gencode.v43.annotation.gff3'\n",
    "#rna_bw_file = '/data/slurm/hejl/riboseq/data/hg38/K562/GSE153597/input_features/rnaseq.bw'\n",
    "rna_bw_file = '/data/slurm/hejl/riboseq/data/hg38/Epithelial/GSE200097/input_features/rnaseq.bw'\n",
    "ribo_bw_file = '/data/slurm/hejl/riboseq/data/hg38/Epithelial/GSE200097/output_features/riboseq.bw'\n",
    "rna_bw_file_t1 = '/data/slurm/hejl/riboseq/data/hg38/Prostate/GSE130465/input_features/rnaseq.bw'\n",
    "rna_bw_file_t2 = '/data/slurm/hejl/riboseq/data/hg38/hTERT-RPE1/GSE138533/input_features/rnaseq.bw'\n",
    "rna_bw_file_t3 = '/data/slurm/hejl/riboseq/data/hg38/RD/GSE103308/input_features/rnaseq.bw'\n",
    "\n",
    "#rna_bw_file = '/data/slurm/hejl/riboseq/data/hg38/mean.sorted.bw'\n",
    "#rna_bw_file = '/data/slurm/hejl/riboseq/Translatomer/data/hg38/mean.sorted.bw'\n",
    "\n",
    "gene_name = 'KLF6'\n",
    "#gene_name = 'KRT19'\n",
    "\n",
    "gff_df = pd.read_csv(gff_file, sep='\\t', comment='#', header=None)\n",
    "gene_info = get_cds_interval(gff_df, gene_name)\n",
    "\n",
    "gene_interval =kipoiseq.Interval(gene_info['chrom'], gene_info['gene_start'], gene_info['gene_end'])\n",
    "region_interval = gene_interval\n",
    "target_interval = Interval(region_interval.chrom, region_interval.start, region_interval.end).resize(SEQUENCE_LENGTH)\n",
    "ref_seq = fasta_extractor.extract(target_interval)\n",
    "ref_emb = torch.Tensor(one_hot_encode(ref_seq)).to(device)\n",
    "\n",
    "epi = torch.Tensor(generate_inputs(region_interval, fasta_file, rna_bw_file)[0]).unsqueeze(1).to(device)\n",
    "reference_input = torch.cat([ref_emb, epi], dim = 1).unsqueeze(0)\n",
    "pred = model(reference_input)[0].detach().cpu().numpy()\n",
    "\n",
    "epi_t1 = torch.Tensor(generate_inputs(region_interval, fasta_file, rna_bw_file_t1)[0]).unsqueeze(1).to(device)\n",
    "reference_input_t1 = torch.cat([ref_emb, epi_t1], dim = 1).unsqueeze(0)\n",
    "pred_t1 = model(reference_input_t1)[0].detach().cpu().numpy()\n",
    "\n",
    "epi_t2 = torch.Tensor(generate_inputs(region_interval, fasta_file, rna_bw_file_t2)[0]).unsqueeze(1).to(device)\n",
    "reference_input_t2 = torch.cat([ref_emb, epi_t2], dim = 1).unsqueeze(0)\n",
    "pred_t2 = model(reference_input_t2)[0].detach().cpu().numpy()\n",
    "\n",
    "epi_t3 = torch.Tensor(generate_inputs(region_interval, fasta_file, rna_bw_file_t3)[0]).unsqueeze(1).to(device)\n",
    "reference_input_t3 = torch.cat([ref_emb, epi_t3], dim = 1).unsqueeze(0)\n",
    "pred_t3 = model(reference_input_t3)[0].detach().cpu().numpy()\n",
    "\n",
    "\n",
    "gt = torch.Tensor(generate_outputs(region_interval, fasta_file, ribo_bw_file)[0]).cpu().numpy()\n",
    "#\n",
    "tracks = {'Ground Truth(Epithelial)': gt,\n",
    "          'Prediction(Epithelial)': pred,          \n",
    "          'Prediction(hTERT-RPE1)': pred_t2,\n",
    "          'Prediction(Prostate)': pred_t1,\n",
    "          'Prediction(RD)': pred_t3,\n",
    "          }\n",
    "\n",
    "plot_tracks(tracks, target_interval, height=1) #\n",
    "\n",
    "correlation, p_value = pearsonr(pred,gt)\n",
    "correlation1, p_value1 = pearsonr(pred_t1,gt)\n",
    "correlation2, p_value2 = pearsonr(pred_t2,gt)\n",
    "correlation3, p_value3 = pearsonr(pred_t3,gt)\n",
    "print('Pearson(Epithelial): ')\n",
    "print(correlation)\n",
    "print('Pearson(hTERT-RPE1): ')\n",
    "print(correlation2)\n",
    "print('Pearson(Prostate): ')\n",
    "print(correlation1)\n",
    "print('Pearson(RD): ')\n",
    "print(correlation3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0809a055",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_tracks(tracks, interval, height=1.5):\n",
    "    fig, axes = plt.subplots(len(tracks), 1, figsize=(20, height * len(tracks)), sharex=True)\n",
    "    for ax, (title, y) in zip(axes, tracks.items()):\n",
    "        ax.fill_between(np.linspace(interval.start, interval.end, num=len(y)), y)\n",
    "        ax.set_title(title)\n",
    "        ax.set_ylim(0, 5)\n",
    "        sns.despine(top=True, right=True, bottom=True)\n",
    "    ax.set_xlabel(\"chr10:3745339-3810875\")\n",
    "    plt.tight_layout()\n",
    "    plt.savefig('/data/slurm/hejl/riboseq/Translatomer/results/Fig4/KLF6.pdf')\n",
    "plot_tracks(tracks, target_interval, height=1)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (pytorch)",
   "language": "python",
   "name": "pytorch"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
